[Prototype] Reload / FileConfigurationProvider

HaoK opened the pull request:
Initial iteration of changes we discussed:
- Moving Reload from ConfigRoot to ConfigProvider
- Introduced new FileConfigurationProvider which contains the reload logic and exposes a single Load(Stream) abstract for xml/json/ini to override

@divega

divega commented on src/Microsoft.Extensions.Configuration.FileProviderExtensions/FileConfigurationProvider.cs:
```diff
@@ -0,0 +1,140 @@
+
+namespace Microsoft.Extensions.Configuration
+{
+    public abstract class FileConfigurationProvider : ConfigurationProvider
```
Assuming taking a dependency on FileProviders is ok we should consider moving this class together with the stuff in Microsoft.Extensions.Configuration.FileExtensions into the same library, then rename it to Microsoft.Extensions.Configuration.FileBased, FileProviders, or something similar.

HaoK replied:
So the current difference in the two package dependencies are:

FileExtensions depends on Abstractions only.
FileProviderExtensions depends on the Configuration implementation and Physical file providers

I would assume this was factored this way for a reason, so we should probably be careful collapsing these together

divega replied:
> I would assume this was factored this way for a reason...

Mainly because different people worked on it (@kirthik and I added FileExtensions, @rynowak and @lodejard added FileProviderExtensions) and that is precisely why we need to refactor now! :smile:

Anyway, I remember there was a reason Physical file provider are considered a heavier/more constraining dependency to have. Maybe @davidfowl or @lodejard can shed some light on it.

divega commented on src/Microsoft.Extensions.Configuration.FileProviderExtensions/FileConfigurationProvider.cs:
```diff
@@ -0,0 +1,140 @@
+            if (fileProvider == null)
+            {
+#if NET451
+                var basePath = AppDomain.CurrentDomain.GetData("APP_CONTEXT_BASE_DIRECTORY") as string ??
```
When/if we merge this with FileExtensions try to have this encapsulated in a private utility method in this library. Bonus points for moving it to a single place where we can all share it :smile:

divega replied:
In fact the actual location of the file should now be computed only once (both for reading the file and for watching it) and take the `ConfigurationBuilder.GetBasePath()` into account.

divega replied:
If I remember correctly the `AddXyz()` methods already call `GetBasePath()` and `Path.Combine()` so the contract is that the path passed to the constructor of any file-based provider is already the full path, so we should:
- Rename the constructor parameter to reflect that a full path is expected
- Change the logic in this method to extract the basePath from the full path of the file.

HaoK replied:
The base path is only used for the PhysicalFileProvider, how about we just set that inside of the AddXyz methods for each provider rather than relying on the provider to compute it later?  So base path won't be recomputed normally, only if you add your own FileConfigProvider and don't specify a IFileProvider

HaoK replied:
Actually looks like we are only using the IFileProvider for the reload watching, and directly using fileStream otherwise.  Maybe we should go through IFileProvider always?

divega replied:
> The base path is only used for the PhysicalFileProvider

Base path is use to find the right file to read and watch and it is actually settable by the user with `configurationBuilder.SetBasePath()` that is why we should only every compute the full path in one place, inside the `AddXyz()` methods.

divega commented on src/Microsoft.Extensions.Configuration.FileProviderExtensions/FileConfigurationProvider.cs:
```diff
@@ -0,0 +1,140 @@
+
+        public abstract void Load(Stream stream);
+
+        public IFileProvider FileProvider { get; set; }
```
It this property ever only used for testing?

HaoK replied:
Right now its only used for testing, but there's nothing preventing it from being a true extensibility point, i.e. if you wanted to replace PhysicalFileProvider with something like a VirtualPathProvider

divega replied:
I am not sure it is needed. If it isn't then I would like to remove it. If it is needed then I think we should rearrange things to that it gets set in product code so that if anyone needs to call the getter it won't always be null.

HaoK commented:
Updated:
- Removed FileProviders project and merged with FileExtensions, kept namespace for FileConfigurationProvider
- Removed Load overload on ConfigurationProvider concrete implementation

HaoK commented:
This will probably be replaced by the outcome of https://github.com/aspnet/Configuration/issues/401

HaoK commented:
Replacing with new PR
